cupy/math/sumprod.py

```diff
@@ -51,12 +51,45 @@ def prod(a, axis=None, dtype=None, out=None, keepdims=False):
 # TODO(okuta): Implement cumprod
 
 
-def cumsum(a, axis=None, dtype=None, out=None):
-    if axis is None:
-        a = a.ravel()
-    else:
-        raise ValueError("'axis' option is not supported")
+def _axis_to_first(x, axis):
+    if axis < 0:
+        axis = x.ndim + axis
+    trans = [axis] + [a for a in range(x.ndim) if a != axis]
+    revert = list(range(1, axis + 1)) + [0] + list(range(axis + 1, x.ndim))
+    return trans, revert
+
+
+def _proc_as_batch(proc, x, axis):
+    trans, revert = _axis_to_first(x, axis)
+    t = x.transpose(trans)
+    s = t.shape
+    r = t.reshape(x.shape[axis], -1).T
+    result = proc(r)
+    return result.T.reshape(s).transpose(revert)
+
+
+def _cumsum_batch(out):
+    kern = core.ElementwiseKernel(
+        'int32 pos, int32 batch', 'raw T x',
+        '''
+        int b = i % batch;
+        int j = i / batch;
+        if (j & pos) {
+          x[(A){b, j}] += x[(A){b, j ^ pos | (pos - 1)}];
```

Comment:
How about making two index variables simply?
```
const int dst_index[] = {b, j};
const int src_index[] = {b, j ^ pos | (pos - 1)};
```

I feel `A` is a little difficult to read.